embed extended header inside .ptd flatbuffer section

JacobSzwejbka opened the pull request:
Summary: Embed the header inside the flatbuffer. We do this for .pte and it lets us reuse a lot of flatbuffer tools natively.

Differential Revision: D68578075

lucylq reviewed (approved):
Lgtm, 

nit to change `extended_header` to `flat_tensor_header` for flat_tensor?

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -109,14 +109,38 @@ runtime::Error save_ptd(
+  out.write(
+      reinterpret_cast<const char*>(builder.GetBufferPointer()) +
+          sizeof(new_offset),
+      4); // This is the 'FT01' magic bytes from flat_tensor.fbs.
```
So this doesn't become stale if "FT01" becomes "FT02"
```suggestion
      4); // This is the file identifier from flat_tensor.fbs.
```

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -109,14 +109,38 @@ runtime::Error save_ptd(
       aligned_size(FlatTensorHeader::kHeaderExpectedLength, tensor_alignment);
 
+  // The general structure of the file is:
+  // [flatbuffer offset to root table][flat tensor magic bytes][header]
```
"flat tensor magic bytes" is ambiguous since it could also refer to the magic in FlatTensorHeader.
"flatbuffer file identifier" would be more specific.

Similar with "header"; could say "FlatTensorHeader". And there may be some padding after the the header before the flatbuffer contents.

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -149,10 +173,10 @@ runtime::Error save_ptd(
           FlatTensorHeader::kHeaderExpectedLength, tensor_alignment));
 
-  // Write flatbuffer
+  // Write flatbuffer, offset by 8 bytes since we wrote those before the header.
```
Some specifics to make it more clear
```suggestion
  // Write flatbuffer, offset by 8 bytes (4-byte root table offset + 4-byte file
  // identifier) since we wrote those before the FlatTensorHeader.
```

dbort reviewed (approved):
Looks good! Thanks for making this change.

A good double-check would be to try parsing a .ptd file using the `flatc` commandline tool, like
```
flatc --json --defaults-json --strict-json -o /tmp extension/flat_tensor/serialize/flat_tensor.fbs -- input.ptd
```
and see if it was able to create /tmp/input.json

JacobSzwejbka committed:
embed extended header inside .ptd flatbuffer section (#7965)

Summary:

Embed the header inside the flatbuffer. We do this for .pte and it lets us reuse a lot of flatbuffer tools natively.

Reviewed By: lucylq

Differential Revision: D68578075
